addons/sourcemod/scripting/freak_fortress_2.sp

```diff
@@ -6543,182 +6553,247 @@ stock GetBossIndex(client)
-	ReplaceString(formula, sizeof(formula), " ", "");  //Get rid of spaces
+	KvGetString(BossKV[Special[client]], key, formula, sizeof(formula), defaultFormula);
+	ReplaceString(formula, sizeof(formula), " ", ""); //Get rid of spaces
 	new length=strlen(formula);
 	for(new i; i<=length; i++)
 	{
 		strcopy(buffer, sizeof(buffer), formula[i]);
-		switch(buffer[0])
+		switch (buffer[0])
 		{
-			case '(':
+			case '\0':
 			{
-				parentheses++;
-				sum[parentheses]=0.0;
-				_operator[parentheses]=0;
+				foundOperator = true; //Probably less hacky way of detecting constant hp.
+				break;
 			}
-			case ')':
+			case '(':
 			{
-				valueReady=true;
-				mustClose=true;
+				if (bracket + 1 == sizeof(_operator))
+				{
+					decl String:bossName[32];
+					KvRewind(BossKV[Special[client]]);
+					KvGetString(BossKV[Special[client]], "name", bossName, sizeof(bossName));
+					LogError("[FF2] %s's %s formula is too complex, using default!", bossName, key);
+					return defaultValue;
+				}
+				closeBracket = false;
+				bracket++;
+				_operator[bracket] = Operation_None;
+				sum[bracket] = 0.0;
+				break;
 			}
-			case '\0':
+			case ')':
 			{
-				valueReady=true;
+				closeBracket = true;
+				break;
 			}
 			case 'n', 'x':
 			{
-				usePlayers=true;
+				usePlayers = true;
+				break;
 			}
 			case '+':
 			{
-				_operator[parentheses]=1;
-				canAdd=true;
+				_operator[bracket] = Operation_Add;
+				foundOperator = true;
+				break;
 			}
 			case '-':
 			{
-				_operator[parentheses]=2;
-				canAdd=true;
+				_operator[bracket] = Operation_Subtract;
+				foundOperator = true;
+				break;
 			}
 			case '*':
 			{
-				_operator[parentheses]=3;
-				canAdd=true;
+				_operator[bracket] = Operation_Multiplication;
+				foundOperator = true;
+				break;
 			}
 			case '/':
 			{
-				_operator[parentheses]=4;
-				canAdd=true;
+				_operator[bracket] = Operation_Division;
+				foundOperator = true;
+				break;
 			}
 			case '^':
 			{
-				_operator[parentheses]=5;
-				canAdd=true;
+				_operator[bracket] = Operation_Exponential;
+				foundOperator = true;
+				break;
 			}
-			default:
+			case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
 			{
+				//It's a number... 100% safer then assuming buffer has a number.
 				StrCat(value, sizeof(value), buffer);
+				break;
+			}
+			default:
+			{
+				decl String:bossName[32];
+				KvRewind(BossKV[Special[client]]);
+				KvGetString(BossKV[Special[client]], "name", bossName, sizeof(bossName));
+				LogError("[FF2] %s's %s formula contains invalid character %c!", bossName, key, buffer[0]);
+				break;
 			}
 		}
-
-		if(valueReady)
+		if (foundOperator || closeBracket)
 		{
-			valueReady=false;
-			if(usePlayers)
+			if (foundOperator)
 			{
-				usePlayers=false;
-				switch(_operator[parentheses])
+				foundOperator = false; //This here is just so I don't forget to tell the code, "We already operated on."
```

Comment:
You wanted a comment, I'm horrible at adding good comments(I'm commentless 99% of the time when I code)